### ISSUE TITLE
Farseek update incompatible with Sponge

### ISSUE THREAD

**Issue by Xplodin** [SpongePowered/SpongeForge#3041]
**I am currently running**

- SpongeForge version: 1.12.2-2838-7.1.8-RC3965.jar
- Forge version: forge-1.12.2-14.23.5.2847-universal.jar
- Java version: 8
- Operating System: Ubuntu

- Plugins/Mods:
AgriCraft (by InfinityRaider)
Ruins (Structure Spawning System) (by atomicstrykergrumpy)
AtomicStryker's Battle Towers (by atomicstrykergrumpy)
Moar Tinkers (by AshuraNoYami)
Waystones (by BlayTheNinth)
IvToolkit (by Ivorforce)
ShadowMC (by ShadowfactsDev)
InstantUnify (by KidsDontPlay)
DimensionalDoors (by DimensionalDevelopment)
Flux Networks (by sonar_sonic)
MysticalLib (by EpicSquid319)
Ding (by ohaiiChun)
Integrated Tunnels (by kroeser)
PlusTiC (by landmaster_phuong0429)
InfinityLib (by InfinityRaider)
iChunUtil (by ohaiiChun)
Mystcraft (by XCompWiz)
Mantle (by mDiyo)
Hats (by ohaiiChun)
Portal Gun (by ohaiiChun)
FTB Quests (by LatvianModder)
Wither Skeleton Tweaks (by Shadows_of_Fire)
Dungeon Tactics (by PegBeard)
Tech Reborn (by modmuss50)
Tinkers' Complement (by KnightMiner)
RFTools (by McJty)
Overloaded Armor Bar (by tfarecnim)
BdLib (by bdew)
Loot Games (by Time_Conqueror)
Energy Converters (by xalcon)
Improved Backpacks (by DreenDexTwitch)
Open Modular Turrets (by Keridos)
Common Capabilities (by kroeser)
AutoRegLib (by Vazkii)
CraftTweaker (by Jaredlll08)
BiblioCraft (by Nuchaz)
Davincis Vessels (by darkevilmac)
TekTopia (by TangoTek)
Immersive Petroleum (by Flaxbeard)
Controlling (by Jaredlll08)
MenuMobs (by Superbas11)
Wawla - What Are We Looking At (by DarkhaxDev)
Thermal Innovation (by TeamCoFH)
Corail Tombstone (by Corail_31)
Enchantment Descriptions (by DarkhaxDev)
ForgeEndertech (by EnderLanky)
FTB Money (by LatvianModder)
Advanced Hook Launchers (by EnderLanky)
Chunk-Pregenerator (by Speiger)
Blur (by tterrag1098)
Instrumental Mobs (by Mrbysco)
Cyclic (by Lothrazar)
Tinker's JEI (by Possible_triangle)
Item Filters (by LatvianModder)
Cooking for Blockheads (by BlayTheNinth)
Biomes O' Plenty (by Forstride)
Just Enough Energistics (JEE) (by TheRealp455w0rd)
Mowzie's Mobs (by bobmowzie)
FTB Utilities (by LatvianModder)
Just Enough Items (JEI) (by mezz)
Auto Ore Dictionary Converter (by MattDahEpic)
MPUtils (by GenDeathrow)
Hardcore Darkness (by Lumien231)
Baubles (by Azanor13)
Mystical World (by EpicSquid319)
Gas Conduits (by pupnewfster)
Tinkers Construct (by mDiyo)
FoamFix for Minecraft (by asiekierka)
FTB Library (by LatvianModder)
Akashic Tome (by Vazkii)
WrapUp (by TheLMiffy1111)
Buildcraft (by Covert_Jaguar)
Sonar Core (by sonar_sonic)
MovingWorld (by darkevilmac)
Applied Energistics 2 (by AlgorithmX2)
Chameleon (by Texelsaur)
Inventory Pets (by Purplicious_Cow_)
Forge MultiPart CBE (by covers1624)
Thermal Cultivation (by TeamCoFH)
Nature's Compass (by Chaosyr)
Simple Storage Network (by Lothrazar)
Roguelike Dungeons (by greymerktv)
Cyclops Core (by kroeser)
CodeChicken Lib 1.8.+ (by covers1624)
Reskillable (by lanse505)
In Control! (by McJty)
Storage Drawers (by Texelsaur)
Mystical Agradditions (by BlakeBr0)
It's the little things (by Zlepper)
Treasure2! (by gottsch)
Simply Jetpacks 2 (by Tomson124)
Redstone Flux (by TeamCoFH)
Just Enough Resources (JER) (by way2muchnoise)
Shadowfacts' Forgelin (by ShadowfactsDev)
Mekanism (by bradyaidanc)
Better Builder's Wands (by Portablejim)
Mekanism Generators (by bradyaidanc)
Mekanism Tools (by bradyaidanc)
Large Ore Deposits (by EnderLanky)
MalisisDoors (by Ordinastie)
Immersive Tech (by FerroO2000)
MalisisCore (by Ordinastie)
Cosmetic Armor Reworked (by LainMI)
HunterIllager (by bagu_chan)
Doomlike Dungeons (by JaredBGreat)
Actually Additions (by Ellpeck)
NotEnoughIDs (by fewizz_)
Farming for Blockheads (by BlayTheNinth)
p455w0rd's Library (by TheRealp455w0rd)
Fish's Undead Rising (by fish0016054)
Wireless Redstone CBE (by covers1624)
Grue (by Shinoow)
Thermal Foundation (by TeamCoFH)
Spice of Life: Carrot Edition (by lordcazsius)
WanionLib (by WanionCane)
Bewitchment (by sunconure11)
MattDahEpic Core (MDECore) (by MattDahEpic)
Construct's Armory (by TheIllusiveC4)
ArchitectureCraft - TridentMC Version (by darkevilmac)
Random Things (by Lumien231)
Ice and Fire: Dragons in a whole new light! (by alex1the1666)
Integrated Dynamics (by kroeser)
NuclearCraft (by tomdodd4598)
OreLib (by OreCruncher)
ConnectedTexturesMod (by tterrag1098)
EnderCore (by tterrag1098)
Bookshelf (by DarkhaxDev)
MineTweaker RecipeMaker (by DoubleDoorDev)
DiscordIntegration (by TheChikachi)
Industrial Foregoing (by Buuz135)
DarknessLib (by Shinoow)
Culinary Construct (by TheIllusiveC4)
Arcane Essentials (Electroblob's Wizardry) (by FavouriteDragon)
Block Drops (JEI Addon) (by KidsDontPlay)
AE2 Stuff (by bdew)
Mob Grinding Utils (by vadis365)
AI Improvements (by DarkGuardsman)
Ore Excavation (by Funwayguy)
Pam's HarvestCraft (by pamharvestcraft)
Thermal Dynamics (by TeamCoFH)
FastWorkbench (by Shadows_of_Fire)
Chisel (by tterrag1098)
Patchouli (by Vazkii)
Paintings ++ (by AbsolemJackdaw)
Farseek (by delvr)
AmbientSounds 3 (by CreativeMD)
Quark (by Vazkii)
Immersive Engineering (by BluSunrize)
Crafting Tweaks (by BlayTheNinth)
Totemic (by ljfa2)
UniDict (by WanionCane)
McJtyLib (by McJty)
Custom Main Menu (by Lumien231)
OMLib (by Keridos)
RandomPatches (by TheRandomLabs)
Chance Cubes (by Turkey2349)
Rats (by alex1the1666)
Streams (by delvr)
Ghost's Explosives (by ghostgamingpe)
TAIGA (Tinkers alloying addon) (by zkafaceTV)
Inventory Tweaks (by Kobata)
AppleSkin (by squeek502)
Resource Loader (by Lumien231)
Dungeons2 (by gottsch)
OldJavaWarning (by DarkhaxDev)
Multi Mob Library (by Daveyx0)
Recurrent Complex (by Ivorforce)
Numina (by machinemuse)
Modular Powersuits (by machinemuse)
Fast Leaf Decay (by olafskiii)
The Summoner (by Daveyx0)
Minecolonies (by H3lay)
Mouse Tweaks (by YaLTeR)
Iron Chests (by ProgWML6)
The Twilight Forest (by Benimatic)
BetterFps (by Guichaguri)
Mystical Agriculture (by BlakeBr0)
Cucumber Library (by BlakeBr0)
Ender IO (by CrazyPants_MC)
LLibrary (by _ForgeUser11902522)
Electroblob's Wizardry (by Electroblob)
Customized Dungeon Loot (by XxRexRaptorxX)
Login Shield (by Glasspelican)
Integration Foregoing (by Jackyy)
LagGoggles (by Terminator_NL)
Level Up! Reloaded (by BeetoGuy)
Primitive Mobs (by Daveyx0)
Stupid Things (by Furgle)
Scannable (by Sangar_)
ICBM - Classic (by DarkGuardsman)
OpenModsLib (by OpenMods)
OpenBlocks (by OpenMods)
Armory Expansion (by AshuraNoYami)
Just Enough Calculation (by towdium)
Pam's Cookables - Pam's Harvestcraft Addon (by BloodWorkXGaming)
Time Core (by Time_Conqueror)
Dynamic Surroundings (by OreCruncher)
[ANGRY PIXEL] The Betweenlands (by oily_oli)
Just Enough HarvestCraft (JEHC) (by mrAppleXZ)
Morpheus (by Quetzi)
Bad Wither No Cookie - Reloaded (by Kreezxil)
Chisels & Bits (by AlgorithmX2)
JourneyMap (by techbrew)
NetherPortalFix (by BlayTheNinth)
Fairy Lights (by pau101)
Hwyla (by TehNut)
GottschCore (by gottsch)
CoFH Core (by TeamCoFH)
Thermal Expansion (by TeamCoFH)
FTB Backups (by LatvianModder)
Tesla Core Lib (by face_of_cat)
Placebo (by Shadows_of_Fire)
Extra Utilities (by RWTema)
LetsEncryptCraft (by CloudyPSP)
Armor Toughness Bar (by tfarecnim)
The Midnight (by cipher_zero_x)
ZeroCore (by ZeroNoRyouki)
Building Gadgets (by Direwolf20)
The Beneath (by Shinoow)
CoFH World (by TeamCoFH)
AppleCore (by squeek502)
/dank/null (by TheRealp455w0rd)
LunatriusCore (by Lunatrius)
More Overlays (by FeldiM245)
JAOPCA (by TheLMiffy1111)
Champions (by TheIllusiveC4)
Waila Harvestability (by squeek502)
Reborn Core (by modmuss50)

**Issue Description**
Wont boot. Just crashes.

https://gist.github.com/Xplodin/5b97ee63b3d46bc564644444cce1fdf4

This worked with other packs on 2847. just not mc.eternal for some reason.
Ive applied all foamfix fixes and other config changes but get the same result

**Comment by clienthax** [comment]
Something is breaking one of our mixins, specifically world.chunk.ChunkMixin_Forge

```
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector redirect$forgeImpl$IgnoreLoadEvent$bam000(Lnet/minecraftforge/fml/common/eventhandler/EventBus;Lnet/minecraftforge/fml/common/eventhandler/Event;)Z in mixins.forge.core.json:world.chunk.ChunkMixin_Forge failed injection check, (0/1) succeeded. Using refmap mixins.forge.refmap.json
```

Can you try without lag goggles, if that doesn't work can you start the server with -Dmixin.debug=true and provide the log.

**Comment by gabizou** [comment]
Can you add `-Dmixin.dumpTargetOnFailure=true` and upload the `.mixin.out` folder after it crashes? There's likely another coremod that is transforming our target (or forge updated their event) and it's causing SpongeForge's target to go missing.

**Comment by Xplodin** [comment]
EDIT: wrong files. uploading new ones

**Comment by Xplodin** [comment]
[Updated latest.log and .mixin.out.zip](https://github.com/SpongePowered/SpongeForge/files/4000905/Updated.latest.log.and.mixin.out.zip)



Sorry about that

**Comment by gabizou** [comment]
**TLDR; Farseek does whacky things and Sponge won't support their way of doing these sort of transformations, because cases like this happen.**

Based on [Farseek doing a wildcard catch all](https://github.com/delvr/Farseek/blob/1.12/src/main/resources/META-INF/farseek_cm.cfg#L4) replacement of `MinecraftForge.EVENT_BUS.post(...)`, all expected replacements of the method are being rewritten to Farseek's package, which doesn't quite make sense.

The two changes that Sponge makes are already laid out in the Mixin in question:
https://github.com/SpongePowered/SpongeForge/blob/ba8e427d60adaf905d5b40e05650809b9c6401c6/src/main/java/org/spongepowered/mod/mixin/core/world/chunk/ChunkMixin_Forge.java#L60-L74

are meant to avoid throwing duplicated events. The problem here is that Sponge doesn't have a way to avoid these sort of counter modifications (in fact, I'd be sure that Farseek is unintentionally redirecting more calls than they are expecting).

Here's the output of the target for Chunk#onLoad (from the mixin output provided by Xplodin (thank you btw, that nailed down the issue 100%):

```java
// Decompiled java output
   public void func_76631_c() {
      this.field_76636_d = true;
      this.field_76637_e.func_147448_a(this.field_150816_i.values());
      ClassInheritanceMultiMap[] var1 = this.field_76645_j;
      int var2 = var1.length;

      for(int var3 = 0; var3 < var2; ++var3) {
         ClassInheritanceMultiMap classinheritancemultimap = var1[var3];
         this.field_76637_e.func_175650_b(ImmutableList.copyOf(classinheritancemultimap));
      }

     // This is where Sponge would be redirecting first
      package.post(MinecraftForge.EVENT_BUS, new Load(this));
   }
```

And the bxytecode output:
```
     public func_76631_c() { //()V
         <localVar:index=4 , name=classinheritancemultimap , desc=Lnet/minecraft/util/ClassInheritanceMultiMap;, sig=Lnet/minecraft/util/ClassInheritanceMultiMap<Lnet/minecraft/entity/Entity;>;, start=L1, end=L2>
         <localVar:index=0 , name=this , desc=Lnet/minecraft/world/chunk/Chunk;, sig=null, start=L3, end=L4>

         L3 {
             aload0 // reference to self
             iconst_1
             putfield net/minecraft/world/chunk/Chunk.field_76636_d:boolean
         }
         L5 {
             aload0 // reference to self
             getfield net/minecraft/world/chunk/Chunk.field_76637_e:net.minecraft.world.World
             aload0 // reference to self
             getfield net/minecraft/world/chunk/Chunk.field_150816_i:java.util.Map
             invokeinterface java/util/Map.values()Ljava/util/Collection;
             invokevirtual net/minecraft/world/World.func_147448_a(Ljava/util/Collection;)V
         }
         L6 {
             aload0 // reference to self
             getfield net/minecraft/world/chunk/Chunk.field_76645_j:net.minecraft.util.ClassInheritanceMultiMap[]
             astore1
             aload1
             arraylength
             istore2
             iconst_0
             istore3
         }
         L7 {
             f_new (Locals[4]: net/minecraft/world/chunk/Chunk, [Lnet/minecraft/util/ClassInheritanceMultiMap;, 1, 1) (Stack[0]: null)
             iload3
             iload2
             if_icmpge L8
             aload1
             iload3
             aaload
             astore4
         }
         L1 {
             aload0 // reference to self
             getfield net/minecraft/world/chunk/Chunk.field_76637_e:net.minecraft.world.World
             aload4
             invokestatic com/google/common/collect/ImmutableList.copyOf(Ljava/util/Collection;)Lcom/google/common/collect/ImmutableList;
             invokevirtual net/minecraft/world/World.func_175650_b(Ljava/util/Collection;)V
         }
         L2 {
             iinc 3 1
             goto L7
         }
         L8 {
             f_new (Locals[1]: net/minecraft/world/chunk/Chunk) (Stack[0]: null)
             getstatic net/minecraftforge/common/MinecraftForge.EVENT_BUS:net.minecraftforge.fml.common.eventhandler.EventBus
             new net/minecraftforge/event/world/ChunkEvent$Load
             dup
             aload0 // reference to self
             invokespecial net/minecraftforge/event/world/ChunkEvent$Load.<init>(Lnet/minecraft/world/chunk/Chunk;)V
             invokestatic farseek/world/gen/structure/package.post(Lnet/minecraftforge/fml/common/eventhandler/EventBus;Lnet/minecraftforge/fml/common/eventhandler/Event;)Z
             pop
         }
         L9 {
             return
         }
         L4 {
         }
     }
```

> I'd be sure that Farseek is unintentionally redirecting more calls than they are expecting

The reason why I say this is that from their transformer, it appears that their transformer is doing a redirect for ALL `EventBus.post()` calls, no matter where they originate from, and this shows in places that they've no business touching, like:

```java
// Biome.java
   public int getWaterColorMultiplier() {
      GetWaterColor event = new GetWaterColor(this, this.field_76759_H);
      package.post(MinecraftForge.EVENT_BUS, event);
      return event.getNewColor();
   }

   public int getModdedBiomeGrassColor(int original) {
      GetGrassColor event = new GetGrassColor(this, original);
      package.post(MinecraftForge.EVENT_BUS, event);
      return event.getNewColor();
   }

   public int getModdedBiomeFoliageColor(int original) {
      GetFoliageColor event = new GetFoliageColor(this, original);
      package.post(MinecraftForge.EVENT_BUS, event);
      return event.getNewColor();
   }
```
```
     public getWaterColorMultiplier() { //()I
         <localVar:index=0 , name=this , desc=Lnet/minecraft/world/biome/Biome;, sig=null, start=L1, end=L2>
         <localVar:index=1 , name=event , desc=Lnet/minecraftforge/event/terraingen/BiomeEvent$GetWaterColor;, sig=null, start=L3, end=L2>

         L1 {
             new net/minecraftforge/event/terraingen/BiomeEvent$GetWaterColor
             dup
             aload0 // reference to self
             aload0 // reference to self
             getfield net/minecraft/world/biome/Biome.field_76759_H:int
             invokespecial net/minecraftforge/event/terraingen/BiomeEvent$GetWaterColor.<init>(Lnet/minecraft/world/biome/Biome;I)V
             astore1
         }
         L3 {
             getstatic net/minecraftforge/common/MinecraftForge.EVENT_BUS:net.minecraftforge.fml.common.eventhandler.EventBus
             aload1
             invokestatic farseek/world/gen/structure/package.post(Lnet/minecraftforge/fml/common/eventhandler/EventBus;Lnet/minecraftforge/fml/common/eventhandler/Event;)Z
             pop
         }
         L4 {
             aload1
             invokevirtual net/minecraftforge/event/terraingen/BiomeEvent$GetWaterColor.getNewColor()I
             ireturn
         }
         L2 {
         }
     }

     public getModdedBiomeGrassColor(int arg0) { //(I)I
         <localVar:index=0 , name=this , desc=Lnet/minecraft/world/biome/Biome;, sig=null, start=L1, end=L2>
         <localVar:index=1 , name=original , desc=I, sig=null, start=L1, end=L2>
         <localVar:index=2 , name=event , desc=Lnet/minecraftforge/event/terraingen/BiomeEvent$GetGrassColor;, sig=null, start=L3, end=L2>

         L1 {
             new net/minecraftforge/event/terraingen/BiomeEvent$GetGrassColor
             dup
             aload0 // reference to self
             iload1
             invokespecial net/minecraftforge/event/terraingen/BiomeEvent$GetGrassColor.<init>(Lnet/minecraft/world/biome/Biome;I)V
             astore2
         }
         L3 {
             getstatic net/minecraftforge/common/MinecraftForge.EVENT_BUS:net.minecraftforge.fml.common.eventhandler.EventBus
             aload2
             invokestatic farseek/world/gen/structure/package.post(Lnet/minecraftforge/fml/common/eventhandler/EventBus;Lnet/minecraftforge/fml/common/eventhandler/Event;)Z
             pop
         }
         L4 {
             aload2
             invokevirtual net/minecraftforge/event/terraingen/BiomeEvent$GetGrassColor.getNewColor()I
             ireturn
         }
         L2 {
         }
     }

     public getModdedBiomeFoliageColor(int arg0) { //(I)I
         <localVar:index=0 , name=this , desc=Lnet/minecraft/world/biome/Biome;, sig=null, start=L1, end=L2>
         <localVar:index=1 , name=original , desc=I, sig=null, start=L1, end=L2>
         <localVar:index=2 , name=event , desc=Lnet/minecraftforge/event/terraingen/BiomeEvent$GetFoliageColor;, sig=null, start=L3, end=L2>

         L1 {
             new net/minecraftforge/event/terraingen/BiomeEvent$GetFoliageColor
             dup
             aload0 // reference to self
             iload1
             invokespecial net/minecraftforge/event/terraingen/BiomeEvent$GetFoliageColor.<init>(Lnet/minecraft/world/biome/Biome;I)V
             astore2
         }
         L3 {
             getstatic net/minecraftforge/common/MinecraftForge.EVENT_BUS:net.minecraftforge.fml.common.eventhandler.EventBus
             aload2
             invokestatic farseek/world/gen/structure/package.post(Lnet/minecraftforge/fml/common/eventhandler/EventBus;Lnet/minecraftforge/fml/common/eventhandler/Event;)Z
             pop
         }
         L4 {
             aload2
             invokevirtual net/minecraftforge/event/terraingen/BiomeEvent$GetFoliageColor.getNewColor()I
             ireturn
         }
         L2 {
         }
     }
```


Since it's been a known transformative issue of how Farseek goes about modifying the game (and Sponge does it's best to guarantee compatibility in a wide variety of configurations), I'll close with a short list of previous issues outlined by Farseek:

https://github.com/SpongePowered/SpongeForge/issues/754
https://github.com/delvr/Farseek/issues/16
